### PR TITLE
Auto-register sonata.page.router

### PR DESCRIPTION
**Pull request by covex-nn**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added auto-registration sonata.page.router to cmf_routing.router service
```

## Subject

Recipe for Flex should not configure other bundles. A package should work with default configuration out from the box. But PageBundle will not work without CmfRouting configuration: `sonata.page.router` service should be added to `cmf_routing.router`:

```yaml
cmf_routing:
    chain:
        routers_by_id:
            sonata.page.router: 150
            router.default: 100
```

This PR allows to auto-register `sonata.page.router` to `cmf_routing.router` with following configuration:

```yaml
sonata_page:
    # this could be a part of recipe for Flex
    router_auto_register:
        enabled: true
        priority: 150
```

This feature is not enabled by default

### CODE REVIEW

File: src/DependencyInjection/Compiler/CmfRouterCompilerPass.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\PageBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+class CmfRouterCompilerPass implements CompilerPassInterface
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/CmfRouterCompilerPass.php]: Should be final

File: src/DependencyInjection/Compiler/CmfRouterCompilerPass.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public function process(ContainerBuilder $container)
+    {
+        $enabled = 'sonata.page.router_auto_register.enabled';
+        if ($container->hasParameter($enabled) && $container->getParameter($enabled)) {
+            $router = $container->getDefinition('cmf_routing.router');
+            $router->addMethodCall('add', [
```
Comment (greg0ire) [inline src/DependencyInjection/Compiler/CmfRouterCompilerPass.php]: You can get rid of this OTV:
`$container->getDefinition('cmf_routing.router')->addMethodCall('add', [`

File: src/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -85,6 +85,17 @@ public function getConfigTreeBuilder()
                     ->thenInvalid('Invalid multisite configuration %s. For more information, see https://sonata-project.org/bundles/page/master/doc/reference/multisite.html')
                 ->end()
             ->end()
+            ->arrayNode('router_auto_register')
+                ->addDefaultsIfNotSet()
+                ->children()
+                    ->booleanNode('enabled')
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: Please add a call to `info`

File: tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                continue;
+            }
+            list($reference, $weight) = $arguments;
+            if ($reference instanceof Reference && 'sonata.page.router' === $reference->__toString()) {
+                if (!$enabled) {
+                    $this->fail('`sonata.page.router` service should not be auto registered');
+                } else {
+                    $this->assertEquals($priority, $weight);
+                    break;
+                }
```
Comment (greg0ire) [inline tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php]: Please avoid negative logic:
```php
if ($enabled) {
    $this->assertEquals($priority, $weight);
    break;
} else {
    $this->fail('`sonata.page.router` service should not be auto registered');
}
$this->fail('`sonata.page.router` service should not be auto registered');
```

File: tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+            if ('add' !== $method) {
+                continue;
+            }
+            list($reference, $weight) = $arguments;
+            if ($reference instanceof Reference && 'sonata.page.router' === $reference->__toString()) {
+                if (!$enabled) {
+                    $this->fail('`sonata.page.router` service should not be auto registered');
+                } else {
+                    $this->assertEquals($priority, $weight);
```
Comment (greg0ire) [inline tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php]: please use `assertSame`

File: tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        foreach ($router->getMethodCalls() as $methodCall) {
+            list($method, $arguments) = $methodCall;
+
+            if ('add' !== $method) {
+                continue;
+            }
+            list($reference, $weight) = $arguments;
+            if ($reference instanceof Reference && 'sonata.page.router' === $reference->__toString()) {
+                if (!$enabled) {
+                    $this->fail('`sonata.page.router` service should not be auto registered');
```
Comment (greg0ire) [inline tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php]: Please use double quotes over backticks

File: tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                }
+            }
+        }
+    }
+
+    public function providerRoutedAutoRegister()
+    {
+        return [
+            [true, 42],
+            [false, 84],
```
Comment (greg0ire) [inline tests/DependencyInjection/Compiler/CmfRouterAutoRegisterTest.php]: Please add key for each set:
`'enabled router' => [true, 42],`

**Comment by covex-nn** [comment]
@greg0ire all done

**Comment by covex-nn** [comment]
@jordisala1991 done, see https://github.com/covex-nn/SonataPageBundle/blob/router-auto-register/docs/reference/installation.rst#cmf-routing-configuration

**Comment by jordisala1991** [comment]
Thank you @covex-nn

**Pull request merged**